p2panda-net/src/sync/accept.rs

```diff
@@ -0,0 +1,221 @@
+                                .into()
+                            )
+                        );
+                    };
+
+                    // From this point on we are only expecting "data" messages from the sync
+                    // session.
+                    let FromSync::Data(header, payload) = message else {
+                        return Err(
+                            SyncError::Critical(
+                                "expected only data messages from sync session in data sync phase"
+                                .into()
+                            )
+                        );
+                    };
+
+                    engine_actor_tx
+                        .send(ToEngineActor::SyncMessage {
+                            header,
+                            payload,
+                            delivered_from: peer,
+                            topic: topic.clone(),
+                        })
+                        .await
+                        .map_err(|err| {
+                            SyncError::Critical(
+                                format!("engine_actor_tx failed sending sync message: {err}")
+                            )
+                        })?;
+                },
+            }
+        }
+
+        // If topic was never set then we didn't receive any messages. In that case, the engine
+        // wasn't ever informed, so we can return here silently.
+        let Some(topic) = topic else {
+            return Ok(());
+        };
+
+        engine_actor_tx
+            .send(ToEngineActor::SyncDone { peer, topic })
+            .await
+            .map_err(|err| {
+                SyncError::Critical(format!("engine_actor_tx failed sending sync done: {err}"))
+            })?;
+
+        Ok(())
+    });
+
+    // Run the "accepting peer" side of the sync protocol.
+    let result = sync_protocol
+        .accept(
+            Box::new(&mut send),
+            Box::new(&mut recv),
+            Box::new(&mut sink),
+        )
+        .await;
+
+    // Drop the tx, so the rx in the glue task receives the closing event.
+    drop(sink);
+
+    // The sync protocol failed and we're informing the "glue" task about it, so it can accordingly
+    // wind down and inform the engine about it.
+    if let Err(sync_session_err) = result {
+        sync_error_tx
+            .send(sync_session_err)
+            .map_err(|err| SyncError::Critical(format!("sync_error_tx failed: {err}")))?;
+    }
+
+    // We're expecting the task to exit with a result soon, we're awaiting it here ..
+    let glue_task_result = glue_task_handle
+        .await
+        .map_err(|err| SyncError::Critical(format!("glue task handle failed: {err}")))?;
+
+    // .. to inform some brave developer who will read the error logs ..
+    if let Err(SyncError::Critical(err)) = &glue_task_result {
+        error!("critical error in sync protocol: {err}");
```

Comment:
There is other places where we print logs on errors, I hope this is not getting too much?

Reply:
In this case the error would be logged twice, right? Once here and once in `src/sync/handler.rs` L66:

`error!("sync failure as acceptor: {}", err)`

I'd prefer just to keep one of them but it's not a blocker for me.

Reply:
Okay, let me remove one of them then 👍🏻